Userland/Libraries/LibJS/Runtime/PromiseJobs.cpp

```diff
@@ -87,77 +78,114 @@ ThrowCompletionOr<Value> PromiseReactionJob::call()
+    auto& handler = reaction.handler();
+    if (handler.has_value()) {
+        // a. Let getHandlerRealmResult be GetFunctionRealm(reaction.[[Handler]].[[Callback]]).
+        auto get_handler_realm_result = get_function_realm(global_object, *handler->callback.cell());
+
+        // b. If getHandlerRealmResult is a normal completion, set handlerRealm to getHandlerRealmResult.[[Value]].
+        if (!get_handler_realm_result.is_throw_completion()) {
+            handler_realm = get_handler_realm_result.release_value();
+        } else {
+            // c. Else, set handlerRealm to the current Realm Record.
+            global_object.vm().clear_exception();
+            handler_realm = global_object.vm().current_realm();
+        }
 
-PromiseResolveThenableJob* PromiseResolveThenableJob::create(GlobalObject& global_object, Promise& promise_to_resolve, Value thenable, JobCallback then)
-{
-    // FIXME: A bunch of stuff regarding realms, see step 2-5 in the spec linked below
-    return global_object.heap().allocate<PromiseResolveThenableJob>(global_object, promise_to_resolve, thenable, then, *global_object.function_prototype());
-}
+        // d. NOTE: handlerRealm is never null unless the handler is undefined. When the handler is a revoked Proxy and no ECMAScript code runs, handlerRealm is used to create error objects.
+    }
 
-PromiseResolveThenableJob::PromiseResolveThenableJob(Promise& promise_to_resolve, Value thenable, JobCallback then, Object& prototype)
-    : NativeFunction(prototype)
-    , m_promise_to_resolve(promise_to_resolve)
-    , m_thenable(thenable)
-    , m_then(then)
-{
+    // 4. Return the Record { [[Job]]: job, [[Realm]]: handlerRealm }.
+    return { move(job), handler_realm };
 }
 
 // 27.2.2.2 NewPromiseResolveThenableJob ( promiseToResolve, thenable, then ), https://tc39.es/ecma262/#sec-newpromiseresolvethenablejob
-ThrowCompletionOr<Value> PromiseResolveThenableJob::call()
+static ThrowCompletionOr<Value> run_resolve_thenable_job(GlobalObject& global_object, Promise& promise_to_resolve, Value thenable, JobCallback& then)
 {
-    auto& vm = this->vm();
-    auto& global_object = this->global_object();
+    auto& vm = global_object.vm();
 
     // a. Let resolvingFunctions be CreateResolvingFunctions(promiseToResolve).
-    auto [resolve_function, reject_function] = m_promise_to_resolve.create_resolving_functions();
+    auto [resolve_function, reject_function] = promise_to_resolve.create_resolving_functions();
 
     // b. Let thenCallResult be HostCallJobCallback(then, thenable, « resolvingFunctions.[[Resolve]], resolvingFunctions.[[Reject]] »).
-    dbgln_if(PROMISE_DEBUG, "[PromiseResolveThenableJob @ {}]: Calling then job callback for thenable {}", this, &m_thenable);
-    auto then_call_result = call_job_callback(global_object, m_then, m_thenable, &resolve_function, &reject_function);
+    dbgln_if(PROMISE_DEBUG, "[PromiseResolveThenableJob]: Calling then job callback for thenable {}", &thenable);
```

Comment:
https://github.com/SerenityOS/serenity/pull/10478#discussion_r732291797

> Since PromiseReactionJob doesn't exist anymore, it should't be referenced in debug output like this either.

Same below.